platform/net/firewall_provider_linux.go

```diff
@@ -3,7 +3,111 @@
+	   second thing in our "BOSH" classid namespace.
+
+	   _Hopefully_ noone uses a major number of "b054", and we avoid collisions _forever_!
+	   If you need to select new classids for firewall rules or traffic control rules, keep
+	   the major number "b054" for bosh stuff, unless there's a good reason to not.
+
+	   The net_cls.classid structure is described in more detail here:
+	   https://www.kernel.org/doc/Documentation/cgroup-v1/net_cls.txt
+	*/
+	natsIsolationClassID uint32 = 2958295042
+)
+
+// SetupNatsFirewall will setup the outgoing cgroup based rule that prevents everything except the agent to open connections to the nats api
 func SetupNatsFirewall(mbus string) error {
+	// return early if
+	// we get a https url for mbus. case for create-env
+	// we get an empty string. case for http_metadata_service (responsible to extract the agent-settings.json from the metadata endpoint)s
+	if mbus == "" || strings.HasPrefix(mbus, "https://") {
+		return nil
+	}
+	mbusURL, err := gonetURL.Parse(mbus)
+	if err != nil || mbusURL.Hostname() == "" {
+		return bosherr.WrapError(err, "Error parsing MbusURL")
+	}
+
+	host, port, err := net.SplitHostPort(mbusURL.Host)
+	if err != nil {
+		return bosherr.WrapError(err, "Error getting Port")
+	}
+	ipt, err := iptables.New()
+	if err != nil {
+		return bosherr.WrapError(err, "Iptables Error")
+	}
+	exists, err := ipt.Exists("mangle", "POSTROUTING",
+		"-d", "127.0.0.1",
+		"-p", "tcp",
+		"--dport", "2822",
+		"-m", "cgroup",
+		"--cgroup", fmt.Sprintf("%v", natsIsolationClassID),
+		"-j", "ACCEPT",
+	)
+	if err != nil {
+		return bosherr.WrapError(err, "Iptables Error checking for monit rule")
+	}
+	if !exists {
+		err = ipt.Insert("mangle", "POSTROUTING", 1,
+			"-d", "127.0.0.1",
+			"-p", "tcp",
+			"--dport", "2822",
+			"-m", "cgroup",
+			"--cgroup", fmt.Sprintf("%v", natsIsolationClassID),
+			"-j", "ACCEPT",
+		)
+		if err != nil {
+			return bosherr.WrapError(err, "Iptables Error inserting for monit rule")
+		}
+	}
+	err = ipt.AppendUnique("mangle", "POSTROUTING",
+		"-d", host,
+		"-p", "tcp",
+		"--dport", port,
+		"-m", "cgroup",
+		"--cgroup", fmt.Sprintf("%v", natsIsolationClassID),
+		"-j", "ACCEPT",
+	)
+	if err != nil {
+		return bosherr.WrapError(err, "Iptables Error inserting for agent ACCEPT rule")
+	}
+	err = ipt.AppendUnique("mangle", "POSTROUTING",
+		"-d", host,
+		"-p", "tcp",
+		"--dport", port,
+		"-j", "DROP",
+	)
+	if err != nil {
+		return bosherr.WrapError(err, "Iptables Error inserting for non-agent DROP rule")
+	}
+	var isolationClassID = natsIsolationClassID
+	natsAPICgroup, err := cgroups.New(cgroups.SingleSubsystem(cgroups.V1, cgroups.NetCLS), cgroups.StaticPath("/nats-api-access"), &specs.LinuxResources{
+		Network: &specs.LinuxNetwork{
+			ClassID: &isolationClassID,
+		},
+	})
+	if err != nil {
+		return bosherr.WrapError(err, "Error setting up cgroups for nats api access")
+	}
 
-	return nil
+	err = natsAPICgroup.AddProc(uint64(os.Getpid()), cgroups.NetCLS)
```

Comment:
The `classid` definition was done before the `iptable` rules are set in the script based version. See https://github.com/cloudfoundry/bosh-linux-stemcell-builder/pull/219/files#diff-9d395104639f75c39a4b71d21f0da02530a19ad79d52dd26e824c43c35e9809cR9. Is it important whether the `classid` definition is configured before or after the `iptable` rules?

Reply:
not to my understanding. basically my understanding of the way this works isL
net_cls tags network packets with the classid for processes in a cgroup. the iptables extension just looks at that tag. so if the cgroup and the associated class id don't exist yet, nothing can have that tag. I don't think that either needs to come first. it's just about when traffic filtering starts.

Reply:
I'm also not so deep into this  that is why I wanted to have a discussion on this.